labcontrol/db/process.py

```diff
@@ -3819,3 +3826,407 @@ def _generate_amplicon_prep_information(self):
+
+                # Note is_control field replaces the line directly below
+                if not d['is_control']:
+                    # strip the prepended study id from orig_name2, but only
+                    # if this is an 'experimental sample' row, and not a
+                    # 'control' row. (captured here w/orig_name2 and study_id
+                    # not equaling None. This also prevents interference w/the
+                    # population of the DataFrame index below, using the
+                    # existing list comprehension.
+                    d['orig_name2'] = re.sub("^%s\." % d['study_id'],
+                                             '',
+                                             d['orig_name2'])
+
+            return results
+
+    def _generate_metagenomics_prep_information(self):
+        """Generates prep information for Metagenomics workflows
+
+        An internal method used to implement the generation of prep information
+        files for Metagenomics workflows. This method is called by
+        generate_prep_information() only. This method is not currently
+        implemented.
+
+        Returns
+        -------
+        dict: { int: str,
+                int: str,
+                int: str,
+                .
+                .
+                .
+                int: str,
+                str: str }
+
+        where 'int: str' represents either a Study ID and a TSV file (in string
+        form), or a Prep ID and TSV file (in string form).
+
+        'str: str' represents controls data; the key is the constant
+        'Controls', and the value is a TSV file (in string form).
+        """
+        results = self._get_metagenomics_data_for_prep()
+
+        data = {}
+
+        for item in results:
+            # format well
+            well = []
+            col = item['col_num']
+            row = item['row_num']
+            while row:
+                row, rem = divmod(row - 1, 26)
+                well[:0] = container_module.LETTERS[rem]
+
+            # adding a new field to the item
+            item['well_id'] = ''.join(well) + str(col)
+
+            # Note: currently we have reverted to generating just one prep
+            # sheet for all items in run, but we anticipate that may change
+            # back in the near future.  That is why we retain the return
+            # structure of a dictionary holding prep sheet strings rather
+            # than returning a single prep sheet string even though, at the
+            # moment, the dictionary will always have only one entry.
+            curr_prep_sheet_id = self.run_name
+            if curr_prep_sheet_id not in data:
+                data[curr_prep_sheet_id] = {}
+
+            content = item['content']
+
+            # adding item to the data (organized by prep_sheet_id and
+            # content string.
+            if content in data[curr_prep_sheet_id]:
+                raise ValueError("'%s' appears more than once" % content)
```

Comment:
~~I believe `content` is a large bolus of text here, which may make digestion of the traceback difficult. Would it be more helpful for an error to dump `curr_prep_sheet_id`? (I'm really not sure)~~

nm, it looks like this is a sample composition? Would it make sense to provide in the traceback both the `curr_prep_sheet_id` and `content`?

Reply:
Not a problem; although currently there is effectively only one prep_sheet_id, due to merging the data from the control and samples files.